*** field.c Wed May 11 09:28:15 2005
--- field.c.new Wed Jan 14 11:44:22 2009
***************
*** 916,922 ****
  
  	parse_field = fw_parse_field;
  	scan = force_string(FIELDWIDTHS_node->var_value)->stptr;
! 	end = scan + 1;
  	if (FIELDWIDTHS == NULL)
  		emalloc(FIELDWIDTHS, int *, fw_alloc * sizeof(int), "set_FIELDWIDTHS");
  	FIELDWIDTHS[0] = 0;
--- 916,922 ----
  
  	parse_field = fw_parse_field;
  	scan = force_string(FIELDWIDTHS_node->var_value)->stptr;
! 
  	if (FIELDWIDTHS == NULL)
  		emalloc(FIELDWIDTHS, int *, fw_alloc * sizeof(int), "set_FIELDWIDTHS");
  	FIELDWIDTHS[0] = 0;
***************
*** 941,947 ****
  		errno = 0;
  		tmp = strtoul(scan, &end, 10);
  		if (errno != 0
! 		    || !(*end == '\0' || is_blank(*end))
  		    || !(0 < tmp && tmp <= INT_MAX))
  			fatal(_("invalid FIELDWIDTHS value, near `%s'"),
  			      scan);
--- 941,947 ----
  		errno = 0;
  		tmp = strtoul(scan, &end, 10);
  		if (errno != 0
! 		    || (*end != '\0' && ! is_blank(*end))
  		    || !(0 < tmp && tmp <= INT_MAX))
  			fatal(_("invalid FIELDWIDTHS value, near `%s'"),
  			      scan);
***************
*** 954,960 ****
  		if (*scan == '\0')
  			break;
  	}
! 	FIELDWIDTHS[i] = -1;
  
  	update_PROCINFO("FS", "FIELDWIDTHS");
  }
--- 954,960 ----
  		if (*scan == '\0')
  			break;
  	}
! 	FIELDWIDTHS[i+1] = -1;
  
  	update_PROCINFO("FS", "FIELDWIDTHS");
  }
